### PR TITLE
Update RGI-4.2.2 wrapper

### PR DESCRIPTION
**Pull request by mjram0s**

### CODE REVIEW

File: rgi.xml
Comment (apetkau) [inline rgi.xml:4]: Since the `rgi` bioconda package already defines all these dependencies, you don't actually have to repeat them in the XML file.  You can remove all but the `rgi` dependency and it should all work.

**Review by raphenya — APPROVED** [review]
Looks good

**Review by raphenya — CHANGES_REQUESTED** [review]
Please change this `Github: https://github.com/phac-nml/rgi` to `Github: https://github.com/arpcard/rgi`

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: For the nucleotide/protein sequences you could probably only leave a few characters (e.g., `ATGC`) which should help shorten this table a bit.

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: So, I believe we want to add `--local` here (and run the `rgi main` command with `--local`).

The reason is, without `--local`, the database is loaded to a single directory relative to where **rgi** is installed (e.g., `/rgi-software/database`). So, if multiple instances of **rgi** are running at the same time, they will all attempt to load the database in the same directory `/rgi-software/database` and could interfere with each other).

However, if you use `--local`, then the database is loaded to a unique directory each time the tool is run (e.g., `/galaxy/job/1/database` for the first execution, `/galaxy/job/2/database` for the second execution and so on).  So, if multiple instances of **rgi** are running at the same time, they won't interfere with each other.

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: I would actually remove this option, since it doesn't produce anything that's noticeable to a user. It can just be defaulted to off (e.g., no `--clean`) which can help with debugging rgi in Galaxy.

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: For parameters that correspond directly to options in the **RGI** command, could you replace `name="input_type"` with `argument="--input_type"`.  You can still use `$input_type` when constructing the command to run above, but this will additionally display the corresponding tool option in the Galaxy UI.  For example:

# rgi options

```
  -t {read,contig,protein,wgs}, --input_type {read,contig,protein,wgs}
                        specify data input type (default = contig)
```

Since there is an option named `--input_type` here, it would make sense to use `argument="--input_type"` below.

# `rgi.xml`

```xml
<param argument="--input_type" type="select" multiple="false" label="Input type">
```

# Galaxy UI

image

This will now display **--input_type** below the tool label, which makes it easier to match up this particular option in the Galaxy UI to the underlying command-line option in **rgi**.

There is a bit more description of all this in a nice guide on best practices for writing Galaxy tools https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#parameter-name-argument-and-help

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: Could you add a `help` attribute here (and in other `<param>` parameters where it makes sense) to describe what the parameter does.  You can get the text from the rgi help section:

# rgi command help

```
  --low_quality         use for short contigs to predict partial genes
```

# `rgi.xml`

```xml
<param name="low_quality" type="boolean" checked="false" truevalue="--low_quality" falsevalue="" label="Low quality" help="use for short contigs to predict partial genes"/>
```

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: Could you surround $input_sequence with single quotes `'$input_sequence'`, which would handle a case where some path to the input file has spaces in it?

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Also, for the orientation, I think  you have to change `+` to `\+` to escape it. Otherwise it gets interpreted as a bullet point:

image

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: Since we're loading the database each time with `rgi load`, there's another option for selecting a database that I think would be useful to include here, especially for testing. This would be allowing users to upload the `card.json` database file to Galaxy and select this file from the history.

You can do this by making use of [conditionals](https://planemo.readthedocs.io/en/latest/writing_standalone.html#conditional-parameters) in Galaxy.  That is, we can have a drop-down menu with two options for selecting the database:

1. Select database managed by Galaxy
2. Select database from history

Selecting **Select database managed by Galaxy** will be the same option we have currently (where you can select a database like `rgi_20181001`).

Selecting **Select database from history** would let users select a JSON file from their history (`card.json`).

An example of what this would look like can be found in the Galaxy **NCBI BLAST+ blastn** tool.

image

Here, there are three options you can select as a source for the database.  Selecting **Locally installed BLAST database** will then display another selection box for pre-installed BLAST databases (e.g., the **NT** database).

image

Selecting **FASTA file from your history** lets you build a BLAST database from a file in your history:

image

Selecting **BLAST database from your history** is similar to the **FASTA** file option, but assumes you have a pre-build BLAST database in your history:

image

The XML file coding this is https://github.com/peterjc/galaxy_blast/blob/master/tools/ncbi_blast_plus/ncbi_macros.xml#L347.

By including the ability to select the CARD/RGI database from your Galaxy history, we could also possibly add automated tests for the tool (since we could include a copy of the `card.json` database file and pass it as input to the tool when testing).

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: I believe you can replace all this with a citation by the **doi** and Galaxy should automatically load up authors/title information.  For example:

```xml
<citation type="doi">10.1093/nar/gkw1004</citation>
```

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: It looks like this file has a mixture of tabs and spaces used for indentation. Could you fix up the file so it uses only one type of character (preferably spaces)?

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks @mjram0s, this looks awesome :smile:. I tested it out and everything works 100%.

I do have some comments for you to improve the Galaxy tool wrapper. If you have any questions please let me know. Same to you @raphenya, if you have any comments on what I've suggested just let me know.

File: rgi.xml
Comment (raphenya) [reply rgi.xml]: Yes, we need `--local`.

File: rgi.xml
Comment (raphenya) [reply rgi.xml]: I think loading `card.json` from history is ideal option

**Review by raphenya — APPROVED** [review]

**Comment by raphenya** [comment]
@mjram0s awesome changes!  @apetkau waiting for your approval, then I will merge this pull request.

File: rgi.xml
Comment (apetkau) [reply rgi.xml:4]: Thank you. This looks great.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks. Looks great.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Looks great.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: It looks like there's still a mixture of spaces and tabs in this file. You may have to enable viewing tabs in the editor you're using to see them.

One command you could use to replace all tabs with spaces is the `sed` command. This will read in a text file, perform operations on it, and print the output back to you (or overwrite the original file).  For example:

```bash
sed -i.bak -e 's/\t/    /g' rgi.xml
```

The following is how to interpret the options:

* `-i.bak`: Perform an operation and overwrite the original file, but first making a backup copy and appending `.bak` to the end of the file name (so you will have a backup named `rgi.xml.bak`).
* `-e 's/\t/    /g'`: The command to run with `sed`
    * `s`: Perform a **s**ubstitution
    * `/\t/`: Find all tab characters
    * `/    /`: Replace all the found tab characters with 4 spaces
    * `/g`: Run this substitute/replace multiple times per line (**g**lobally). Without `g` only the first tab is replaced by spaces in each line.

After running this command you may have to take a look at the file and verify that the XML tags still line up.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: This is awesome. Thanks @mjram0s. Works great.

File: rgi.xml
Comment (apetkau) [inline rgi.xml]: Could you surround `$db_opts.rgi_db_hist` by quotes?

File: rgi.xml
Comment (mjram0s) [reply rgi.xml]: These lines have been deleted in the latest commit.

**Comment by apetkau** [comment]
One other comment I have @mjram0s, you have included test files under `test-data`, but there are no `<test>` cases included. There's some examples of how to include test cases in https://planemo.readthedocs.io/en/latest/writing_advanced.html#test-driven-development. Could you include a test case (which just runs RGI with one of the test files and verifies it gets something as output that makes sense)?

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks.

File: rgi.xml
Comment (apetkau) [reply rgi.xml]: Thanks

**Comment by apetkau** [comment]
This looks great to me. Thanks @mjram0s. @raphenya you can merge this.

**Review by apetkau — APPROVED** [review]

**Comment by raphenya** [comment]
@mjram0s @apetkau Thank you.

**Pull request merged**
